metadata-ingestion-modules/airflow-plugin/tests/integration/test_plugin.py

```diff
@@ -0,0 +1,389 @@
+import contextlib
+import dataclasses
+import functools
+import logging
+import os
+import pathlib
+import random
+import signal
+import subprocess
+import time
+from typing import Iterator, Sequence
+
+import pytest
+import requests
+import tenacity
+from airflow.models.connection import Connection
+from datahub.testing.compare_metadata_json import assert_metadata_files_equal
+
+from datahub_airflow_plugin._airflow_shims import (
+    HAS_AIRFLOW_DAG_LISTENER_API,
+    HAS_AIRFLOW_LISTENER_API,
+    HAS_AIRFLOW_STANDALONE_CMD,
+)
+
+pytestmark = pytest.mark.integration
+
+logger = logging.getLogger(__name__)
+IS_LOCAL = os.environ.get("CI", "false") == "false"
+
+DAGS_FOLDER = pathlib.Path(__file__).parent / "dags"
+GOLDENS_FOLDER = pathlib.Path(__file__).parent / "goldens"
+
+
+@dataclasses.dataclass
+class AirflowInstance:
+    airflow_home: pathlib.Path
+    airflow_port: int
+    pid: int
+    env_vars: dict
+
+    username: str
+    password: str
+
+    metadata_file: pathlib.Path
+
+    @property
+    def airflow_url(self) -> str:
+        return f"http://localhost:{self.airflow_port}"
+
+    @functools.cached_property
+    def session(self) -> requests.Session:
+        session = requests.Session()
+        session.auth = (self.username, self.password)
+        return session
+
+
+@tenacity.retry(
+    reraise=True,
+    wait=tenacity.wait_fixed(1),
+    stop=tenacity.stop_after_delay(60),
+    retry=tenacity.retry_if_exception_type(
+        (AssertionError, requests.exceptions.RequestException)
+    ),
+)
+def _wait_for_airflow_healthy(airflow_port: int) -> None:
+    print("Checking if Airflow is ready...")
+    res = requests.get(f"http://localhost:{airflow_port}/health", timeout=5)
+    res.raise_for_status()
+
+    airflow_health = res.json()
+    assert airflow_health["metadatabase"]["status"] == "healthy"
+    assert airflow_health["scheduler"]["status"] == "healthy"
+
+
+class NotReadyError(Exception):
+    pass
+
+
+@tenacity.retry(
+    reraise=True,
+    wait=tenacity.wait_fixed(1),
+    stop=tenacity.stop_after_delay(60),
+    retry=tenacity.retry_if_exception_type(NotReadyError),
+)
+def _wait_for_dag_finish(
+    airflow_instance: AirflowInstance, dag_id: str, require_success: bool
+) -> None:
+    print("Checking if DAG is finished")
+    res = airflow_instance.session.get(
+        f"{airflow_instance.airflow_url}/api/v1/dags/{dag_id}/dagRuns", timeout=5
```

Comment:
If this timeout is reached, do we not retry via tenacity, because exception type won't be `NotReadyError`? Is this intended?

Reply:
at this point the API should be up. if the API is not accessible, we want to fail immediately